addressmanager/src/lib.rs

```diff
@@ -0,0 +1,200 @@
+        Self { ip, port }
+    }
+}
+
+impl AddressManager {
+    pub fn new(db: Arc<DB>) -> Arc<Mutex<Self>> {
+        Arc::new(Mutex::new(Self {
+            banned_address_store: DbBannedAddressesStore::new(db.clone(), MAX_ADDRESSES as u64),
+            not_banned_address_store: not_banned_address_store_with_cache::new(db),
+        }))
+    }
+
+    pub fn add_address(&mut self, address: NetAddress) {
+        // TODO: Don't add non routable addresses
+
+        // We mark `connectionFailedCount` as 0 only after first success
+        self.not_banned_address_store.set(address, 1);
+    }
+
+    pub fn mark_connection_failure(&mut self, address: NetAddress) {
+        let new_count = self.not_banned_address_store.get(address) + 1;
+        if new_count > MAX_CONNECTION_FAILED_COUNT {
+            self.not_banned_address_store.remove(address);
+        } else {
+            self.not_banned_address_store.set(address, new_count);
+        }
+    }
+
+    pub fn mark_connection_success(&mut self, address: NetAddress) {
+        self.not_banned_address_store.set(address, 0);
+    }
+
+    pub fn get_all_addresses(&self) -> impl Iterator<Item = NetAddress> + '_ {
+        self.not_banned_address_store.get_all_addresses()
+    }
+
+    pub fn get_random_addresses(&self, count: usize, exceptions: HashSet<NetAddress>) -> Vec<NetAddress> {
+        self.not_banned_address_store.get_random_addresses(count, exceptions)
+    }
+
+    pub fn ban(&mut self, ip: Ipv6Addr) {
+        self.banned_address_store.set(ip, unix_time()).unwrap();
+        self.not_banned_address_store.remove_by_ip(ip);
+    }
+
+    pub fn unban(&mut self, ip: Ipv6Addr) {
+        self.banned_address_store.remove(ip).unwrap();
+    }
+
+    pub fn is_banned(&mut self, ip: Ipv6Addr) -> bool {
+        const MAX_BANNED_TIME: u64 = 24 * 60 * 60 * 1000;
+        match self.banned_address_store.get(ip).unwrap_option() {
+            Some(timestamp) => {
+                if unix_time() - timestamp > MAX_BANNED_TIME {
+                    self.unban(ip);
+                    false
+                } else {
+                    true
+                }
+            }
+            None => false,
+        }
+    }
+}
+
+fn unix_time() -> u64 {
+    SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as u64
+}
+
+mod not_banned_address_store_with_cache {
+    use std::{
+        cmp::min,
+        collections::{HashMap, HashSet},
+        net::Ipv6Addr,
+        sync::Arc,
+    };
+
+    use database::db::DB;
+    use itertools::Itertools;
+    use rand::{distributions::WeightedIndex, prelude::Distribution};
+
+    use crate::{
+        stores::not_banned_address_store::{DbNotBannedAddressesStore, NotBannedAddressesStore, NotBannedAddressesStoreReader},
+        NetAddress, MAX_ADDRESSES, MAX_CONNECTION_FAILED_COUNT,
+    };
+
+    pub struct Store {
+        db_store: DbNotBannedAddressesStore,
+        addresses: HashMap<NetAddress, u64>,
```

Comment:
I think we should have a newtype around the u64 to avoid errors with the banned u64 which represents time: 
```pub struct ConnectionFailureCount(pub u64)``` 

and of course use it in the DB store as well